include/crm/common/pacemakerd_types.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2004-2019 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU Lesser General Public License
+ * version 2.1 or later (LGPLv2.1+) WITHOUT ANY WARRANTY.
+ */
+
+#ifndef PACEMAKERD_TYPES__H
+#  define PACEMAKERD_TYPES__H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <time.h>
+
+enum pacemakerd_conn_state {
+    pacemakerd_conn_connected,
+    pacemakerd_conn_disconnected
+};
+
+enum pacemakerd_state {
+    pacemakerd_state_invalid = -1,
+    pacemakerd_state_init = 0,
+    pacemakerd_state_starting_daemons,
+    pacemakerd_state_wait_for_ping,
+    pacemakerd_state_running,
+    pacemakerd_state_shutting_down,
+    pacemakerd_state_shutdown_complete,
+    pacemakerd_state_max = pacemakerd_state_shutdown_complete,
+};
+
+typedef struct pacemakerd_s pacemakerd_t;
+
+typedef struct pacemakerd_api_operations_s {
+    int (*connect) (pacemakerd_t *pacemakerd, const char *name);
+    int (*disconnect) (pacemakerd_t *pacemakerd);
+    void (*free) (pacemakerd_t *pacemakerd);
+    int (*set_ping_callback) (pacemakerd_t *pacemakerd,
+                              void (*callback) (pacemakerd_t *pacemakerd,
+                                                time_t last_good,
+                                                enum pacemakerd_state state,
+                                                int rc, gpointer userdata),
+                              gpointer userdata);
+    int (*set_disconnect_callback) (pacemakerd_t *pacemakerd,
+                                    void (*callback) (gpointer userdata),
+                                    gpointer userdata
+                                    );
+    int (*ping) (pacemakerd_t *pacemakerd, const char *name,
+                 const char *admin_uuid, int call_options);
+} pacemakerd_api_operations_t;
+
+struct pacemakerd_s {
+    enum pacemakerd_conn_state conn_state;
+    void *pacemakerd_private;
+    pacemakerd_api_operations_t *cmds;
```

Comment:
Do we really want to break with that? crm_ipc_t is rather the exception so far while other APIs don't make much use of virtual methods either.
Getters and setters aren't optimized away as when used with C++ - right. But of course the impact is probably minor compared to the flexibility they offer over direct access.
Not seeing that we are actually using the possibilities so far but there are probably a lot of things that could be done with methods called via pointers.
Things coming to my mind is handling of deprecated stuff like on the fly insertion of a wrapper, add a new prototype of a method at the end while keeping the deprecated original with an altered name at the old place so that things compiled against the old types will still work while stuff compiled at the new versions will see the new prototype - all without version-handling and stuff.

Reply:
I don't see the value for a case like this, but I'm fine with keeping pointers if you want them. I'd still hide conn_state (callers shouldn't change it), and I'd put the pointers in the main struct rather than use a separate struct for them (since we're hiding everything else anyway).

Reply:
If we decide to walk away from the pointers on the longer run and with new APIs for now I have nothing against it. Just wanted to be sure we aren't overlooking some potential of what we already have before we alter it or proceed in an inconsistent way.
Having the connection-state visible was a decision for consistency and a little bit of performance. Was as well not sure if it could make sense to put a timestamp into it as well having some kind of default ping-handler that just updates this timestamp and does the
pinging/reconnecting stuff autonomously. But of course we can read that timestamp with a
getter as well - with single-threaded pacemaker we probably won't need a wrapper for atomic-access but who knows what the future will bring.
Of course there is no need for having the pointers in an extra allocated memory once we don't
want to grow the data in the base-structure and the function-pointer-list both in a compatible manner any more.

Reply:
Ah, I didn't think of being able to add to the bottom of each separately. Though even then, it would just be for visual organization -- you could mix new members and methods at the end of a single struct.

I like hiding the entire type partly because we have complete freedom over the implementation (don't have to add things at the end, can change types of existing members, etc.) and partly because it's impossible for callers to misuse it by statically declaring a struct or using sizeof() on it. Using the void* private technique gets most of that though.